Threadpool improvements

CookiePLMonster opened the pull request:
This PR consists of several independent improvements to `WorkerThread`, `ThreadPool` and `GlobalThreadPool` classes. Those include:

- proper thread safety for `GlobalThreadPool`.
- interface cleanup for `WorkerThread`.
- fixes for spurious wakeups in `WorkerThread`, which also allowed me to retire blocking while constructing such, dubiously synced using a volatile bool.
- cleaned up `ThreadPool` worker threads init, now starting as many threads as **actually** used. Previously, the last thread would never be used as the callee also counts as performing work.

About the last listed change - I made it start one thread less, but maybe it should still start it and actually use it, by dividing work between `numThreads_ + 1` threads?

CookiePLMonster committed:
Clean up ThreadPool and WorkerThread constructors with C++11 features,

removes variable instantiation from constructor initializer lists

CookiePLMonster committed:
Use move semantics for worker thread functors

CookiePLMonster committed:
Handle spurious wakeups on worker thread condition variables properly

CookiePLMonster committed:
Remove 'started' bool from worker threads, as fixes for spurious wakeups made it obsolete

CookiePLMonster committed:
Cleanup mutex usage in worker threads

CookiePLMonster committed:
Store std::thread in worker thread class by value and explicitly start it up, and don't use shared_ptr in thread pools

CookiePLMonster committed:
Start as many worker threads as actually used (used to start more)

CookiePLMonster committed:
Init GlobalThreadPool in a thread safe manner

hrydgard commented:
Gonna look at it tomorrow but this sounds great! Been meaning to improve the threadpool and use it for more things.

CookiePLMonster committed:
Add MakeUnique.h so make_unique is available without C++14

Includes make_unique from C++14 and make_unique_default_init from C++20,
because it doesn't require compiler support

CookiePLMonster commented:
I have retrofitted `MakeUnique.h`, a new file in `Common` which implements `make_unique` and `make_unique_default_init` for non-C++14 platforms.

I have implemented it basing on docs from here:
https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique
and based SFINAE overload resolution on `memory` from Visual Studio 2017, adapting it to use only C++11 features.

This should make the PR compileable on all platforms again.

unknownbrackets reviewed (commented):
MakeUnique.h needs to be in CMakeLists.txt to show up in IDEs properly.

That said, it's for two usages which both have trivial arguments.  I'd personally just use `std::unique_ptr<ThreadPool>(new ThreadPool(g_Config.iNumWorkerThreads));`.  The syntactic sugar here seems pretty minor...

-[Unknown]

hrydgard commented:
I agree with unknown that MakeUnique.h is not worth it. Other than that, looks great!

CookiePLMonster commented:
I'm not going to exceessively argue about it, but since PPSSPP is using `std::make_shared` happily where available I would say that allowing `make_unique` to be used freely brings extra consistency to the codebase. I can of course remove it, but I would say there is no reason not to use it now that it's been done (and it always has been, in `armips` 3rdparty).

hrydgard commented:
PPSSPP is generally quite inconsistent in coding styles, due to its long life and the evolution of both our programming styles and the C++ language itself in the meantime. make_shared is by no means consistently used. But okay, I accept that argument.

unknownbrackets commented:
(it should still be added to CMakeLists.txt...)

-[Unknown]
